File: backend/app/api/v1/routers/tasks.py
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, Query, Response, UploadFile, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, require_roles
from app.db.session import get_db
from app.models.user import User
from app.repositories.task_repository import TaskRepository
from app.repositories.user_repository import UserRepository
from app.schemas.auth import MessageResponse
from app.schemas.task import (
    TaskDetailRead,
    TaskListResponse,
    TaskMediaRead,
    TaskMessageCreate,
    TaskMessageRead,
    TaskModuleSettingsRead,
    TaskModuleSettingsUpdate,
    TaskUpdateRequest,
)
from app.services.task_service import TaskService

router = APIRouter()


def _service(db: Session) -> TaskService:
    return TaskService(TaskRepository(db), UserRepository(db))


@router.get("/module", response_model=TaskModuleSettingsRead)
def get_module_settings(
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(get_current_user)],
) -> TaskModuleSettingsRead:
    return _service(db).get_module_settings()


@router.patch("/module", response_model=TaskModuleSettingsRead)
def update_module_settings(
    payload: TaskModuleSettingsUpdate,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_roles("admin"))],
) -> TaskModuleSettingsRead:
    return _service(db).update_module_settings(payload.is_active)


@router.get("", response_model=TaskListResponse)
def list_tasks(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin", "employee"))],
    search: Annotated[str | None, Query()] = None,
    status_filter: Annotated[str | None, Query(alias="status")] = None,
    created_from: Annotated[datetime | None, Query()] = None,
    created_to: Annotated[datetime | None, Query()] = None,
    modified_from: Annotated[datetime | None, Query()] = None,
    modified_to: Annotated[datetime | None, Query()] = None,
    sort: Annotated[str | None, Query()] = None,
) -> TaskListResponse:
    return _service(db).list_tasks(
        current_user,
        search,
        status_filter,
        created_from,
        created_to,
        modified_from,
        modified_to,
        sort,
    )


@router.post("", response_model=TaskDetailRead, status_code=status.HTTP_201_CREATED)
async def create_task(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin"))],
    name: Annotated[str, Form(alias="name")],
    description: Annotated[str | None, Form(alias="description")] = None,
    initial_status: Annotated[str, Form(alias="initial_status")] = "todo",
    assigned_user_ids: Annotated[str | None, Form(alias="assigned_user_ids")] = None,
    cover_photo: Annotated[UploadFile | None, File(alias="cover_photo")] = None,
) -> TaskDetailRead:
    cover_data = await cover_photo.read() if cover_photo is not None else None
    return _service(db).create_task(
        user=current_user,
        name=name,
        description=description,
        status_value=initial_status,
        assigned_user_ids=TaskService.parse_assigned_user_ids(assigned_user_ids),
        cover_photo_name=cover_photo.filename if cover_photo is not None else None,
        cover_photo_mime=cover_photo.content_type if cover_photo is not None else None,
        cover_photo_data=cover_data,
    )


@router.get("/{task_id}", response_model=TaskDetailRead)
def get_task(
    task_id: int,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_roles("admin", "employee"))],
) -> TaskDetailRead:
    return _service(db).get_task_detail(task_id)


@router.patch("/{task_id}", response_model=TaskDetailRead)
def update_task(
    task_id: int,
    payload: TaskUpdateRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin", "employee"))],
) -> TaskDetailRead:
    return _service(db).update_task(current_user, task_id, payload)


@router.delete("/{task_id}", response_model=MessageResponse)
def delete_task(
    task_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin"))],
) -> MessageResponse:
    _service(db).delete_task(current_user, task_id)
    return MessageResponse(message="Task deleted")


@router.post("/{task_id}/messages", response_model=TaskMessageRead, status_code=status.HTTP_201_CREATED)
def add_task_message(
    task_id: int,
    payload: TaskMessageCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin", "employee"))],
) -> TaskMessageRead:
    return _service(db).add_message(current_user, task_id, payload.content)


@router.post("/{task_id}/media", response_model=TaskMediaRead, status_code=status.HTTP_201_CREATED)
async def add_task_media(
    task_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(require_roles("admin"))],
    file: Annotated[UploadFile, File(alias="file")],
) -> TaskMediaRead:
    payload = await file.read()
    return _service(db).add_media(
        current_user,
        task_id,
        file.filename or "attachment",
        file.content_type or "application/octet-stream",
        payload,
    )


@router.get("/{task_id}/cover")
def get_task_cover(
    task_id: int,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_roles("admin", "employee"))],
) -> Response:
    file_name, media_type, data = _service(db).get_cover_photo(task_id)
    return Response(content=data, media_type=media_type, headers={"Content-Disposition": f'inline; filename="{file_name}"'})


@router.get("/{task_id}/media/{media_id}")
def get_task_media(
    task_id: int,
    media_id: int,
    db: Annotated[Session, Depends(get_db)],
    _: Annotated[User, Depends(require_roles("admin", "employee"))],
) -> Response:
    file_name, media_type, data = _service(db).get_task_media(task_id, media_id)
    return Response(content=data, media_type=media_type, headers={"Content-Disposition": f'inline; filename="{file_name}"'})

File: frontend/src/services/tasks.ts
import { api } from "./api";

export type TaskStatus = "todo" | "in_progress" | "done";

export type TaskAssignee = {
  id: number;
  name: string;
  email: string;
  job_title: string | null;
  is_active: boolean;
};

export type TaskCard = {
  id: number;
  code: string;
  name: string;
  description: string | null;
  status: TaskStatus;
  created_at: string;
  updated_at: string;
  has_cover_photo: boolean;
  assignees: TaskAssignee[];
};

export type TaskMessage = {
  id: number;
  sender: TaskAssignee;
  content: string;
  created_at: string;
};

export type TaskMedia = {
  id: number;
  file_name: string;
  file_mime: string;
  created_at: string;
  uploaded_by: TaskAssignee;
};

export type TaskDetail = TaskCard & {
  media: TaskMedia[];
  messages: TaskMessage[];
};

export type TaskModuleSettings = {
  is_active: boolean;
};

export type EmployeeRecord = {
  id: number;
  name: string;
  email: string;
  is_active: boolean;
  role: string;
  job_title: string | null;
  profile_photo_name?: string | null;
  created_at: string;
  updated_at: string;
};

export type TaskFilters = {
  search?: string;
  status?: "all" | TaskStatus;
  created_from?: string;
  created_to?: string;
  modified_from?: string;
  modified_to?: string;
  sort?: "created_desc" | "created_asc" | "name_asc" | "name_desc";
};

export const tasksService = {
  async getModuleSettings() {
    const { data } = await api.get<TaskModuleSettings>("/tasks/module");
    return data;
  },

  async updateModuleSettings(isActive: boolean) {
    const { data } = await api.patch<TaskModuleSettings>("/tasks/module", { is_active: isActive });
    return data;
  },

  async listTasks(filters: TaskFilters) {
    const params: Record<string, string> = {};
    if (filters.search) params.search = filters.search;
    if (filters.status && filters.status !== "all") params.status = filters.status;
    if (filters.created_from) params.created_from = `${filters.created_from}T00:00:00`;
    if (filters.created_to) params.created_to = `${filters.created_to}T23:59:59`;
    if (filters.modified_from) params.modified_from = `${filters.modified_from}T00:00:00`;
    if (filters.modified_to) params.modified_to = `${filters.modified_to}T23:59:59`;
    if (filters.sort) params.sort = filters.sort;

    const { data } = await api.get<{ items: TaskCard[] }>("/tasks", { params });
    return data.items;
  },

  async getTask(taskId: number) {
    const { data } = await api.get<TaskDetail>(`/tasks/${taskId}`);
    return data;
  },

  async createTask(payload: {
    name: string;
    description?: string;
    initialStatus: TaskStatus;
    assignedUserIds: number[];
    coverPhoto?: File | null;
  }) {
    const formData = new FormData();
    formData.append("name", payload.name);
    formData.append("description", payload.description ?? "");
    formData.append("initial_status", payload.initialStatus);
    formData.append("assigned_user_ids", JSON.stringify(payload.assignedUserIds));
    if (payload.coverPhoto) {
      formData.append("cover_photo", payload.coverPhoto);
    }
    const { data } = await api.post<TaskDetail>("/tasks", formData, {
      headers: { "Content-Type": "multipart/form-data" }
    });
    return data;
  },

  async updateTask(taskId: number, payload: { name?: string; description?: string | null; status?: TaskStatus; assigned_user_ids?: number[] }) {
    const { data } = await api.patch<TaskDetail>(`/tasks/${taskId}`, payload);
    return data;
  },

  async deleteTask(taskId: number) {
    await api.delete(`/tasks/${taskId}`);
  },

  async addMessage(taskId: number, content: string) {
    const { data } = await api.post<TaskMessage>(`/tasks/${taskId}/messages`, { content });
    return data;
  },

  async addMedia(taskId: number, file: File) {
    const formData = new FormData();
    formData.append("file", file);
    const { data } = await api.post<TaskMedia>(`/tasks/${taskId}/media`, formData, {
      headers: { "Content-Type": "multipart/form-data" }
    });
    return data;
  },

  async getCoverPhoto(taskId: number) {
    const { data } = await api.get<Blob>(`/tasks/${taskId}/cover`, { responseType: "blob" });
    return data;
  },

  async getMedia(taskId: number, mediaId: number) {
    const { data, headers } = await api.get<Blob>(`/tasks/${taskId}/media/${mediaId}`, { responseType: "blob" });
    return {
      blob: data,
      contentType: headers["content-type"] as string | undefined
    };
  },

  async listEmployees() {
    const { data } = await api.get<EmployeeRecord[]>("/users/employees");
    return data;
  },

  async createEmployee(payload: {
    fullName: string;
    email: string;
    password: string;
    jobTitle?: string;
    profilePhoto?: File | null;
  }) {
    const formData = new FormData();
    formData.append("full_name", payload.fullName);
    formData.append("email", payload.email);
    formData.append("password", payload.password);
    formData.append("job_title", payload.jobTitle ?? "");
    if (payload.profilePhoto) {
      formData.append("profile_photo", payload.profilePhoto);
    }
    const { data } = await api.post<EmployeeRecord>("/users/employees", formData, {
      headers: { "Content-Type": "multipart/form-data" }
    });
    return data;
  },

  async updateEmployee(
    employeeId: number,
    payload: {
      fullName?: string;
      email?: string;
      password?: string;
      jobTitle?: string;
      isActive?: boolean;
      profilePhoto?: File | null;
    }
  ) {
    const formData = new FormData();
    if (payload.fullName !== undefined) formData.append("full_name", payload.fullName);
    if (payload.email !== undefined) formData.append("email", payload.email);
    if (payload.password) formData.append("password", payload.password);
    if (payload.jobTitle !== undefined) formData.append("job_title", payload.jobTitle);
    if (payload.isActive !== undefined) formData.append("is_active", String(payload.isActive));
    if (payload.profilePhoto) formData.append("profile_photo", payload.profilePhoto);
    const { data } = await api.patch<EmployeeRecord>(`/users/employees/${employeeId}`, formData, {
      headers: { "Content-Type": "multipart/form-data" }
    });
    return data;
  }
};
